test/Validator/AuthenticationTest.php

```diff
@@ -106,4 +122,97 @@ public function testWithContext()
         $this->assertEquals('myusername', $adapter->getIdentity());
         $this->assertEquals('mypassword', $adapter->getCredential());
     }
+
+    /**
+     * @param int   $code
+     * @param bool  $valid
+     * @param array $messages
+     *
+     * @dataProvider errorMessagesProvider
+     */
+    public function testErrorMessages($code, $valid, $messages)
+    {
+        $adapter = new AuthTest\TestAsset\ValidatableAdapter($code);
+
+        $this->validator->setAdapter($adapter);
+        $this->validator->setService($this->authService);
+        $this->validator->setIdentity('username');
+        $this->validator->setCredential('credential');
+
+        $this->assertEquals($valid, $this->validator->isValid());
+        $this->assertEquals($messages, $this->validator->getMessages());
+    }
+
+    /**
+     * Test using Authentication Service's adapter
+     */
+    public function testUsingAdapterFromService()
+    {
+        $this->authService->setAdapter($this->authAdapter);
+
+        $this->validator->setService($this->authService);
+        $this->validator->setIdentity('username');
+        $this->validator->isValid('password');
+
+        $this->assertEquals('username', $this->validator->getIdentity());
+        $this->assertEquals('password', $this->validator->getCredential());
+        $this->assertEquals('username', $this->authAdapter->getIdentity());
+        $this->assertEquals('password', $this->authAdapter->getCredential());
+        $this->assertNull($this->validator->getAdapter());
+        $this->assertTrue($this->validator->isValid());
+    }
+
+    /**
+     * Ensures that isValid() throws an exception when Authentication Service's
+     * adapter is not an instance of ValidatableAdapterInterface
+     */
+    public function testUsingNoValiatableAdapterFromServiceThrowsRuntimeException()
+    {
+        $this->setExpectedException(
+            'RuntimeException',
+            'Adapter must be an instance of ValidatableAdapterInterface, ' .
+            'ZendTest\Authentication\TestAsset\SuccessAdapter given'
+        );
+
+        $adapter = new AuthTest\TestAsset\SuccessAdapter();
+        $this->authService->setAdapter($adapter);
+
+        $this->validator->setService($this->authService);
+        $this->validator->setIdentity('username');
+        $this->validator->isValid('password');
+    }
+
+    public function errorMessagesProvider()
+    {
+        return [
+            [
```

Comment:
Provide names for each data set; these help us locate the specific set that caused an error more easily.

I'll do this in a separate commit that I push to your branch.